Support layer LSTM

Vimalakirti opened the pull request:
This PR makes the following changes for proving and verifying an LSTM:
- `onnx.rs`: support multiple outputs from a layer and filter out optional input tags (which are "")
- Add layers and basicblocks for `ConstantOfShape`, `Split`, `LSTM`, `Tanh`, `Sigmoid`

But currently this PR only supports LSTM with `batch_size=1` and `seq_len=1`.

(LSTM layer has been successfully proven and verified. The onnx for testing the LSTM layer is in https://drive.google.com/file/d/13RsN1x5m5JGd0PvBBSSVbOzNF8gzGB2q/view?usp=drive_link)

Vimalakirti committed:
add concat, squeeze, and unsqueeze layers

Vimalakirti committed:
issues addressed

Vimalakirti committed:
handle special cases and add test for concat

Vimalakirti committed:
support lstm with seqlen 1

Vimalakirti committed:
cleanup

Vimalakirti committed:
add range, ceil, slice, split, scatternd

Vimalakirti committed:
address comments for concat

liliatangxy commented on src/basic_block/split.rs:
```diff
@@ -0,0 +1,62 @@
+impl BasicBlock for SplitBasicBlock {
+  fn run(&self, _model: &ArrayD<Fr>, inputs: &Vec<&ArrayD<Fr>>) -> Vec<ArrayD<Fr>> {
+    assert!(inputs.len() == 1);
+    assert!(self.axis != inputs[0].ndim() - 1);
```
I think we need to change `!=` to `<`

liliatangxy commented on src/basic_block/split.rs:
```diff
@@ -0,0 +1,62 @@
+use ndarray::{ArrayD, Axis};
+use rand::rngs::StdRng;
+
+// only support concat over dim 0 for now
```
I think this supports `self.axis < inputs[0].ndim() - 1`

liliatangxy commented on src/onnx.rs:
```diff
@@ -152,14 +160,16 @@ pub fn load_file(filename: &str) -> (Graph, Vec<ArrayD<Fr>>) {
     let start_idx = graph.nodes.len() as i32;
     for local_node in local_graph.nodes.iter() {
+      // filter out node input that are ""
+      let node_input = &node.input.iter().filter(|x| x as &str != "").collect::<Vec<_>>();
```
What node inputs are `""`?

Vimalakirti replied:
I remember it's a Slice node in RNNT. An optional field that is supposed to be named "steps" (or something else) turns out to be named "", and results in graph compilation errors. Because "" only happens when the field value is not provided, I think we can directly filter it out to prevent possible errors.

Vimalakirti committed:
Merge branch 'master' into lstm

Vimalakirti committed:
fix automatic check errors

Vimalakirti committed:
add layers for lstm

Vimalakirti committed:
remove ceil

Vimalakirti committed:
cleanup

Vimalakirti committed:
cargo fmt

Vimalakirti commented:
This PR is now ready for reviews. It introduces several new layers for proving LSTM and modifies some old ones:
- Ceil, Sigmoid, Tanh: implemented this by nonlinear macro
- ConstantOfShape: implemented ConstOfShapeBasicBlock
- LSTM: now it supports arbitrary sequence length
- Range: the range check relies on CQ
- ScatterND: by CopyConstraint
- Slice: by CopyConstraint
- Split:  when split along the last axis, we perform Permute first. (like reshape layer)
- Equal: modified old one and performed nonzero check with CQ (modified the setup field of CQ for this purpose)

Besides, some files are modified because of formatting.

Vimalakirti committed:
make it clear

liliatangxy commented on src/basic_block/cq.rs:
```diff
@@ -16,12 +16,13 @@ use std::collections::HashMap;
 #[derive(Debug)]
 pub struct CQBasicBlock {
-  pub setup: Option<(i32, usize)>,
+  pub setup: ArrayD<Fr>,
```
Is this used to support non-contiguous ranges? I think we can make the type be a 1d array or a vec

Vimalakirti committed:
simplify some sublayers in lstm

liliatangxy commented on src/basic_block/cq.rs:
```diff
@@ -16,12 +16,13 @@ use std::collections::HashMap;
   fn genModel(&self) -> ArrayD<Fr> {
-    Array1::from_iter(self.setup.unwrap().0..self.setup.unwrap().0 + (self.setup.unwrap().1 as i32)).map(|x| Fr::from(*x)).into_dyn()
+    // Array1::from_iter(self.setup.unwrap().0..self.setup.unwrap().0 + (self.setup.unwrap().1 as i32)).map(|x| Fr::from(*x)).into_dyn()
+    self.setup.clone()
```
remove comment

liliatangxy commented on src/layer/lstm.rs:
```diff
@@ -0,0 +1,310 @@
+pub struct LSTMLayer;
+impl Layer for LSTMLayer {
+  fn graph(input_shapes: &Vec<&Vec<usize>>, _constants: &Vec<Option<&ArrayD<Fr>>>, _attributes: &Vec<&AttributeProto>) -> (Graph, Vec<Vec<usize>>) {
+    // currently, we do not support P (peepholes) and different sequence lengths
```
is the comment up to date

Vimalakirti replied:
yes, it's up to date. We do not support peepholes and do not support different sequence lengths in a batch. I removed "different sequence lengths" because we require batch_size=1 now.

liliatangxy commented on src/layer/lstm.rs:
```diff
@@ -0,0 +1,310 @@
+    let R_T_output = R_index;
+
+    // sublayer 4: Split for B
+    // Here, we need to transpose B first to split it because we cannot split along the last axis
```
I think for now this is ok but we should raise an issue for proving this case of Split since it is currently annoying to do with permute/copy constraint due to producing multiple outputs

Vimalakirti replied:
The issue is created here: 
https://github.com/uiuc-kang-lab/zk-torch/issues/54

liliatangxy commented on src/layer/lstm.rs:
```diff
@@ -0,0 +1,310 @@
+      N: 1,
+    }));
+    let add_output = graph.addNode(add, vec![(B_split_output[0], 0), (B_split_output[1], 0)]);
+    let sublayer5 = add_output;
```
is there a reason to have these as two separate lines? readability-wise i think it's ok without since the node var name and the comment make it clear

Vimalakirti committed:
update cq setup to arr1

Vimalakirti committed:
update split and lstm

Vimalakirti committed:
further simplify lstm

ddkang commented on src/basic_block/constant.rs:
```diff
@@ -40,3 +40,14 @@ impl BasicBlock for Const2BasicBlock {
 }
+
+#[derive(Debug)]
+pub struct ConstOfShapeBasicBlock {
```
I don't know what this is

Vimalakirti replied:
Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

ddkang commented on src/basic_block/eq.rs:
```diff
@@ -41,3 +42,28 @@ impl BasicBlock for EqBasicBlock {
+  fn run(&self, _model: &ArrayD<Fr>, inputs: &Vec<&ArrayD<Fr>>) -> Vec<ArrayD<Fr>> {
+    assert!(inputs.len() == 2 && inputs[0].ndim() <= 1 && inputs[1].ndim() <= 1);
+    let mut r = ArrayD::zeros(IxDyn(&[std::cmp::max(inputs[0].len(), inputs[1].len())]));
+    if inputs[0].len() == 1 && inputs[1].ndim() > 0 {
```
Comments

Vimalakirti replied:
Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

ddkang commented on src/basic_block/range.rs:
```diff
@@ -0,0 +1,22 @@
+  pub limit: Fr,
+  pub delta: Fr,
+}
+impl BasicBlock for RangeBasicBlock {
```
Does this need to be proved to be secure?

Vimalakirti replied:
Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

ddkang commented on src/layer/arithmetic.rs:
```diff
@@ -8,24 +8,24 @@ use tract_onnx::pb::AttributeProto;
 macro_rules! define_arithmetic_layer {
   ($struct_name:ident, $basic_block:ident) => {
-      pub struct $struct_name;
+    pub struct $struct_name;
```
Why was this file changed?

Vimalakirti replied:
It changed because of formatting. The contents were not changed.

ddkang commented on src/layer/constantofshape.rs:
```diff
@@ -0,0 +1,25 @@
+use tract_onnx::pb::AttributeProto;
+
+pub struct ConstOfShapeLayer;
+impl Layer for ConstOfShapeLayer {
```
Don't know what this does

Vimalakirti replied:
Responded here:
https://github.com/uiuc-kang-lab/zk-torch/pull/57#pullrequestreview-2176836918

ddkang commented on src/layer/equal.rs:
```diff
@@ -1,15 +1,15 @@
 impl Layer for EqualLayer {
-  fn graph(input_shapes: &Vec<&Vec<usize>>, constants: &Vec<Option<&ArrayD<Fr>>>, _attributes: &Vec<&AttributeProto>) -> (Graph, Vec<Vec<usize>>) {
-    assert!(*constants[1].unwrap().first().unwrap() == Fr::from(0));
+  fn graph(input_shapes: &Vec<&Vec<usize>>, _constants: &Vec<Option<&ArrayD<Fr>>>, _attributes: &Vec<&AttributeProto>) -> (Graph, Vec<Vec<usize>>) {
```
Why did this get so much more complex?

ddkang replied:
It seems we should special case some easy cases for high performance

Vimalakirti replied:
The original EqualLayer returns A==1, and it requires all elements in A must be 1 or 0. The updated EqualLayer returns A==B, which completely implements the ONNX Equal. 
I agree considering simple cases can improve our performance, the issue is created here:
https://github.com/uiuc-kang-lab/zk-torch/issues/56

ddkang reviewed (commented):
Please wait for my review before merging in the future, especially for large PRs like this
